armi/nuclearDataIO/cccc/dif3d.py

```diff
@@ -0,0 +1,237 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Module for reading from and writing to DIF3D files, which are module dependent
+binary inputs for the DIF3D code.
+"""
+
+from armi import runLog
+from armi.nuclearDataIO import cccc
+
+
+FILE_SPEC_2D_PARAMS = (
+    [
+        "IPROBT",
+        "ISOLNT",
+        "IXTRAP",
+        "MINBSZ",
+        "NOUTMX",
+        "IRSTRT",
+        "LIMTIM",
+        "NUPMAX",
+        "IOSAVE",
+        "IOMEG1",
+        "INRMAX",
+        "NUMORP",
+        "IRETRN",
+    ]
+    + [f"IEDF{e}" for e in range(1, 11)]
+    + [
+        "NOUTBQ",
+        "I0FLUX",
+        "NOEDIT",
+        "NOD3ED",
+        "ISRHED",
+        "NSN",
+        "NSWMAX",
+        "NAPRX",
+        "NAPRXZ",
+        "NFMCMX",
+        "NXYSWP",
+        "NZSWP",
+        "ISYMF",
+        "NCMRZS",
+        "ISEXTR",
+        "NPNO",
+        "NXTR",
+        "IOMEG2",
+        "IFULL",
+        "NVFLAG",
+        "ISIMPL",
+        "IWNHFL",
+        "IPERT",
+        "IHARM",
+    ]
+)
+
+FILE_SPEC_3D_PARAMS = [
+    "EPS1",
+    "EPS2",
+    "EPS3",
+    "EFFK",
+    "FISMIN",
+    "PSINRM",
+    "POWIN",
+    "SIGBAR",
+    "EFFKQ",
+    "EPSWP",
+] + [f"DUM{e}" for e in range(1, 21)]
+
+
+class Dif3dData(cccc.DataContainer):
+    def __init__(self):
+        cccc.DataContainer.__init__(self)
+
+        self.twoD = {e: None for e in FILE_SPEC_2D_PARAMS}
+        self.threeD = {e: None for e in FILE_SPEC_3D_PARAMS}
+        self.fourD = None
+        self.fiveD = None
+
+
+class Dif3dStream(cccc.StreamWithDataContainer):
+    @staticmethod
+    def _getDataContainer() -> Dif3dData:
+        return Dif3dData()
+
+    def _rwFileID(self) -> None:
+        """
+        Record for file identification information.
+
+        The parameters are stored as a dictionary under the attribute `metadata`.
+
+        Parameters
+        ----------
+        None
+        """
+        with self.createRecord() as record:
+            for param in ["HNAME", "HUSE1", "HUSE2"]:
+                self._metadata[param] = record.rwString(self._metadata[param], 8)
+            self._metadata["VERSION"] = record.rwInt(self._metadata["VERSION"])
+
+    def _rw1DRecord(self) -> None:
+        """
+        Record for problem title, storage, and dump specifications.
+
+        The parameters are stored as a dictionary under the attribute `metadata`.
+
+        Parameters
+        ----------
+        None
+        """
+        with self.createRecord() as record:
+            for i in range(11):
```

Comment:
`11` seems magical to me.  I assume `10` must be part of the file spec.

Maybe `10` should be a const at the top of the file?  Maybe the same with `"MAXSIZ"` and the like.

Maybe not.  

_Normally_, it's good practice to pull all these "magic numbers" out into a place where they can be commented on an explained.  Maybe it's obvious here that these are part of the file spec, and less explanation is needed.

I don't know. Your call.

Reply:
I'm fine playing it safe. done!